crates/prover-services/src/parallel/new.rs

```diff
@@ -0,0 +1,329 @@
+where
+    Da: DaService,
+    Vm: ZkvmHost,
+    Stf: StateTransitionFunction<Vm::Guest, Da::Spec> + Send + Sync,
+    Stf::PreState: Clone + Send + Sync,
+{
+    thread_pool: rayon::ThreadPool,
+
+    proof_mode: Arc<Mutex<ProofGenMode<Da, Vm, Stf>>>,
+
+    da_service: Da,
+    vm: Vm,
+    zk_storage: Stf::PreState,
+    ledger_db: LedgerDB,
+
+    current_da_hash: Option<<Da::Spec as DaSpec>::SlotHash>,
+    proof_queue: Vec<ProofData>,
+}
+
+impl<Da, Vm, Stf> ParallelProverService<Da, Vm, Stf>
+where
+    Da: DaService,
+    Vm: ZkvmHost + 'static,
+    Stf: StateTransitionFunction<Vm::Guest, Da::Spec> + Send + Sync + 'static,
+    Stf::PreState: Clone + Send + Sync,
+{
+    /// Creates a new prover.
+    pub fn new(
+        da_service: Da,
+        vm: Vm,
+        proof_mode: ProofGenMode<Da, Vm, Stf>,
+        zk_storage: Stf::PreState,
+        thread_pool_size: usize,
+        ledger_db: LedgerDB,
+    ) -> anyhow::Result<Self> {
+        match proof_mode {
+            ProofGenMode::Skip => {
+                tracing::info!("Prover is configured to skip proving");
+            }
+            ProofGenMode::Simulate(_) => {
+                tracing::info!("Prover is configured to simulate proving");
+            }
+            ProofGenMode::Execute => {
+                tracing::info!("Prover is configured to execute proving");
+            }
+            ProofGenMode::Prove => {
+                tracing::info!("Prover is configured to prove");
+            }
+        };
+
+        let thread_pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(thread_pool_size)
+            .build()
+            .expect("Thread pool must be built");
+
+        Ok(Self {
+            thread_pool,
+            proof_mode: Arc::new(Mutex::new(proof_mode)),
+            da_service,
+            vm,
+            zk_storage,
+            ledger_db,
+            current_da_hash: None,
+            proof_queue: vec![],
+        })
+    }
+
+    /// Creates a new `ParallelProverService` with the default number of thread pool size.
+    /// Default thread pool size is num_cpus - 1.
+    pub fn new_with_default_workers(
+        da_service: Da,
+        vm: Vm,
+        proof_mode: ProofGenMode<Da, Vm, Stf>,
+        zk_storage: Stf::PreState,
+        ledger_db: LedgerDB,
+    ) -> anyhow::Result<Self> {
+        let num_cpus = num_cpus::get();
+        assert!(
+            num_cpus > 1,
+            "Parallel prover service requires at least 2 available cores to run smoothly"
+        );
+
+        Self::new(
+            da_service,
+            vm,
+            proof_mode,
+            zk_storage,
+            num_cpus - 1,
+            ledger_db,
+        )
+    }
+
+    pub fn set_current_da_hash(&mut self, da_hash: <Da::Spec as DaSpec>::SlotHash) {
+        assert!(
+            self.current_da_hash.is_none(),
+            "Da hash to prove should never be set twice"
+        );
+        self.current_da_hash = Some(da_hash);
```

Comment:
if we are taking the queue order of the proofs as absolute order why would we need the da hash in prover srevice?

Reply:
we dont. removed